src/util/submit-addon.js

```diff
@@ -0,0 +1,388 @@
+    const dest = `${this.downloadDir}/${filename}`;
+    const response = await this.fetch(fileUrl);
+    if (!response.ok || !response.body) {
+      this.logger.log(`Download of signed xpi failed: ${response.status}.`);
+      throw new Error(`Downloading ${filename} failed`);
+    }
+    await this.saveToFile(response.body, dest);
+    return Promise.resolve({
+      success: true,
+      id: addonId,
+      downloadedFiles: [filename],
+    });
+  }
+
+  async postNewAddon(
+    xpiPath: string,
+    channel: string,
+    metaDataJson: Object
+  ): Promise<SignResult> {
+    const upload_uuid = await this.doUploadSubmit(xpiPath, channel);
+
+    const versionObject =
+      channel === 'listed' ? 'current_version' : 'latest_unlisted_version';
+    const { guid: addonId, [versionObject]: {id: newVersionId}} =
+      await this.doNewAddonSubmit(upload_uuid, metaDataJson);
+
+    const fileUrl = await this.waitForApproval(addonId, newVersionId);
+
+    return this.downloadSignedFile(fileUrl, addonId);
+  }
+
+  async putVersion(
+    xpiPath: string,
+    channel: string,
+    addonId: string,
+    metaDataJson: Object
+  ): Promise<SignResult> {
+    const upload_uuid = await this.doUploadSubmit(xpiPath, channel);
+
+    await this.doNewAddonOrVersionSubmit(addonId, upload_uuid, metaDataJson);
+
+    const url =
+      `${this.apiUrlPrefix}addon/${addonId}/versions/?filter=all_with_unlisted`;
+    const {results: [{id: newVersionId}]} =
+      await this.getJson(await this.fetch(url));
+
+    const fileUrl = await this.waitForApproval(addonId, newVersionId);
+
+    return this.downloadSignedFile(fileUrl, addonId);
+  }
+}
+
+type signAddonParams = {|
+  apiKey: string,
+  apiSecret: string,
+  apiUrlPrefix: string,
+  // apiProxy,
+  timeout: number,
+  // verbose,
+  id?: string,
+  xpiPath: string,
+  downloadDir: string,
+  channel: string,
+|}
+
+export async function signAddon({
+  apiKey,
+  apiSecret,
+  apiUrlPrefix,
+  // apiProxy,
+  timeout,
+  // verbose,
+  id,
+  xpiPath,
+  downloadDir,
+  channel,
+}: signAddonParams): Promise<SignResult> {
+  function reportEmpty(name) {
+    throw new Error(`required argument was empty: ${name}`);
+  }
+
+  if (!xpiPath) {
+    reportEmpty('xpiPath');
+  }
+
+  if (!apiSecret) {
+    reportEmpty('apiSecret');
+  }
+
+  if (!apiKey) {
+    reportEmpty('apiKey');
+  }
+
+  try {
+    const stats = await fsPromises.stat(xpiPath);
+
+    if (!stats.isFile()) {
+      throw new Error(`not a file: ${xpiPath}`);
+    }
+  } catch (statError) {
+    throw new Error(`error with ${xpiPath}: ${statError}`);
+  }
+
+  const client = new Client({
+    apiKey,
+    apiSecret,
+    apiUrlPrefix,
+    // apiProxy,
+    validationCheckTimeout: timeout,
+    approvalCheckTimeout: timeout,
+    // verbose,
+    downloadDir,
+  });
+
+  try {
+    if (id === undefined) {
+      return client.postNewAddon(
+        xpiPath,
+        channel,
+        {},
+      );
+    } else {
```

Comment:
I see that then there is no id being passed, we consider it a new addon to be created, can we actually hit this from the `web-ext sign` command or is that meant to be covered by a separate followup?

This part of the PR description  comment `a) web-ext doesn't have a way of the user specifying "this is a new add-on" vs. "this is a new version" ` makes me think it may be something expected to be exposed in a followup, is that the case or are that part of the pull request description comment and this code actually unrelated?

Reply:
yes, it can be reached if the developer passes in an `id` parameter to the `web-ext sign` command, or if there is an id saved in `.web-extension-id` in the `sourceDir`.  But the id would also need to be in the manifest too or they'll be an error (thrown here, locally, because otherwise the API itself would 400)

> This part of the PR description comment a) web-ext doesn't have a way of the user specifying "this is a new add-on" vs. "this is a new version" makes me think it may be something expected to be exposed in a followup, is that the case or are that part of the pull request description comment and this code actually unrelated?

it's related... but the part your quoting is background explaining why the `/addons/addon/<slug>/versions/` endpoint isn't used, rather than anything that I was expecting to change in a follow-up.  I can just delete everything inside the brackets if that's clearer.